### PR TITLE
extract the logic that yields sets of required length to a separate funciton, and test all its edge cases explicitly

### PR DESCRIPTION
**Pull request by AlexCue987**
extract the logic that yields sets of required length to a separate funciton, and test all its edge cases explicitly

### CODE REVIEW

File: kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/sets.kt [hunk 1]
```diff
@@ -45,25 +45,30 @@ fun <A> Arb.Companion.set(gen: Gen<A>, range: IntRange = PropertyTesting.default
    return arbitrary(SetShrinker(range)) {
       val genIter = gen.generate(it).iterator()
       val targetSize = it.random.nextInt(range)
-      val set = mutableSetOf<A>()
       var iterations = 0
```
[the remark this answers is not in the document]
Comment (AlexCue987) [reply kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/sets.kt]: done

File: kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/sets.kt [hunk 2]
```diff
@@ -45,25 +45,30 @@ fun <A> Arb.Companion.set(gen: Gen<A>, range: IntRange = PropertyTesting.default
...
       // The slippage factor controls how many times we will accept a non unique element before giving up,
       // which is the number of elements in the target set * slippage
-      while (iterations < maxMisses && set.size < targetSize && genIter.hasNext()) {
-         val size = set.size
-         set.add(genIter.next().value)
-         if (set.size == size) iterations++
-      }
+      val set = takeSet(elements, targetSize, maxMisses)
       check(set.size == targetSize) {
          "the target size requirement of $targetSize could not be satisfied after $iterations consecutive samples"
```
[the remark this answers is not in the document]
Comment (AlexCue987) [reply kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/sets.kt]: fixed

File: kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/sets.kt [hunk 3]
```diff
@@ -45,25 +45,30 @@ fun <A> Arb.Companion.set(gen: Gen<A>, range: IntRange = PropertyTesting.default
...
    }
 }
 
+internal fun<A> takeSet(elements: Sequence<A>, targetSize: Int, maxMisses: Int): Set<A> {
+   var elementsTaken = 0
+   val chunk = elements
+      .takeWhile { elementsTaken++ < maxMisses }
+      .distinct()
+      .take(targetSize)
+   return chunk.toSet()
```
[the remark this answers is not in the document]
Comment (AlexCue987) [reply kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/sets.kt:68]: replacing idiomatic Kotlin with a while loop? no thanks

File: kotest-property/src/jvmTest/kotlin/com/sksamuel/kotest/property/arbitrary/SetsTest.kt [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+class SetsTest: StringSpec() {
+   init {
+      "generate sets of required length" {
+         checkAll(1_000, Arb.set(
+            Arb.double(),
+            range = 3..4,
+            slippage = 10,
+            )) {
+            val actual = it
+            actual.size.shouldBeIn(3, 4)
```
[the remark this answers is not in the document]
Comment (AlexCue987) [reply kotest-property/src/jvmTest/kotlin/com/sksamuel/kotest/property/arbitrary/SetsTest.kt]: done

File: kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/sets.kt [hunk 5]
```diff
@@ -45,25 +45,29 @@ fun <A> Arb.Companion.set(gen: Gen<A>, range: IntRange = PropertyTesting.default
...
+      val set = takeSet(elements, targetSize, maxAttempts)
       check(set.size == targetSize) {
-         "the target size requirement of $targetSize could not be satisfied after $iterations consecutive samples"
+         "the target size requirement of $targetSize could not be satisfied"
       }
       set
    }
 }
 
+internal fun<A> takeSet(elements: Sequence<A>, targetSize: Int, maxAttempts: Int): Set<A> {
```
Comment (alfonsoristorato) [inline kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/sets.kt:62]: is this internal for testing purposes only? asking cause it will be exposed in the public api

File: kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/sets.kt [hunk 6]
```diff
@@ -45,25 +45,29 @@ fun <A> Arb.Companion.set(gen: Gen<A>, range: IntRange = PropertyTesting.default
...
+   var elementsTaken = 0
+   val chunk = elements
+      .takeWhile { elementsTaken++ < maxAttempts }
+      .distinct()
```
Comment (alfonsoristorato) [inline kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/sets.kt:66]: before we were taking care of dupes in the loop, while now we rely on `distinct` - do you think it will cause memory problems for users that generate big enough data?

**Review by alfonsoristorato — COMMENTED** [review]
little comments / questions

File: kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/sets.kt (hunk 5 above)
Comment (AlexCue987) [reply kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/sets.kt:62]: > is this internal for testing purposes only? asking cause it will be exposed in the public api

it's internal because I did not intend to expose it. why should I?

File: kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/sets.kt (hunk 6 above)
Comment (AlexCue987) [reply kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/sets.kt:66]: > before we were taking care of dupes in the loop, while now we rely on `distinct` - do you think it will cause memory problems for users that generate big enough data?

why write bespoke code when `distinct` already does it for you, is well-tested, etc

File: kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/sets.kt (hunk 5 above)
Comment (alfonsoristorato) [reply kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/sets.kt:62]: I assumed that internal functions go into the public API drop, I was suggesting at making it private. Unless you need it callable for some other reason or some other classes

File: kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/sets.kt (hunk 6 above)
Comment (alfonsoristorato) [reply kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/sets.kt:66]: No reason, I just wondered if this was going to be less memory efficient, cause the way I understand this is that the distinct will create another loop to do the check. But happy either way

**Review by alfonsoristorato — APPROVED** [review]

File: kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/sets.kt (hunk 5 above)
Comment (AlexCue987) [reply kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/sets.kt:62]: > I assumed that internal functions go into the public API drop, I was suggesting at making it private. Unless you need it callable for some other reason or some other classes

if it's private, I cannot explicilty unit test it, right? if it's internal, I can, and it won't generate into API dump. Please correct me if I am wrong.

File: kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/sets.kt (hunk 6 above)
Comment (AlexCue987) [reply kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/sets.kt:66]: > No reason, I just wondered if this was going to be less memory efficient, cause the way I understand this is that the distinct will create another loop to do the check. But happy either way

I concur that `distinct` builds another set and then throws it away, and as such it uses more memory. But I was more concerned about maintainability, and I thought that later on code built from standard well-known blocks would be easier to grok than the bespoke while loop.

File: kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/sets.kt (hunk 5 above)
Comment (alfonsoristorato) [reply kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/sets.kt:62]: I assumed internal functions get exposed in the API, but clearly they don't, all good :)
